arcgis-ios-sdk-samples/Edit data/Edit with branch versioning/EditWithBranchVersioningViewController.swift

```diff
@@ -0,0 +1,542 @@
+            // If the text field is empty, do nothing.
+            guard let branchText = branchTextField.text?.trimmingCharacters(in: .whitespacesAndNewlines), !branchText.isEmpty else { return }
+            let descriptionText = descriptionTextField.text
+            let parameters = self.createServiceParameters(accessPermission: permission, uniqueName: branchText, description: descriptionText)
+            completion(parameters)
+        }
+        createAction.isEnabled = false
+        alertController.addAction(cancelAction)
+        alertController.addAction(createAction)
+        alertController.preferredAction = createAction
+        
+        // The text field for version name.
+        alertController.addTextField { textField in
+            textField.placeholder = "Version name must be unique"
+            textField.delegate = self
+            textFieldObserver = NotificationCenter.default.addObserver(
+                forName: UITextField.textDidChangeNotification,
+                object: textField,
+                queue: .main
+            ) { _ in
+                let text = alertController.textFields?[0].text?.trimmingCharacters(in: .whitespacesAndNewlines)
+                // Enable the done button if
+                // 1. branch version textfield is not empty.
+                let notEmpty = text?.isEmpty == false
+                // 2. branch version string does not exceed 62 characters.
+                let noLongerThan62Characters = text?.count ?? 0 <= 62
+                createAction.isEnabled = notEmpty && noLongerThan62Characters
+            }
+        }
+        // The text field for version description.
+        alertController.addTextField { textField in
+            textField.placeholder = "Branch version description here"
+        }
+        present(alertController, animated: true)
+    }
+    
+    func editFeatureDamageAttribute(feature: AGSFeature, sourceCallout: AGSCallout) {
+        let alertController = UIAlertController(
+            title: "Choose a damage type for the building",
+            message: nil,
+            preferredStyle: .actionSheet
+        )
+        DamageType.allCases.forEach { type in
+            let action = UIAlertAction(title: type.rawValue, style: .default) { _ in
+                feature.attributes["typdamage"] = type.rawValue
+                feature.featureTable?.update(feature)
+                self.mapView.callout.dismiss()
+            }
+            alertController.addAction(action)
+        }
+        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel) { _ in
+            self.mapView.callout.dismiss()
+        }
+        alertController.addAction(cancelAction)
+        let sourceRect = CGRect(origin: CGPoint(x: sourceCallout.bounds.maxX, y: sourceCallout.bounds.midY), size: .zero)
+        alertController.popoverPresentationController?.sourceView = sourceCallout
+        alertController.popoverPresentationController?.sourceRect = sourceRect
```

Comment:
> ...since it isn't even visible when the popover is presented.

This popover is shown when the accessory button in a callout is tapped, so it is guaranteed that popover presents when a callout is showing.

> Instead, the source rect should be the rect of the feature and the source view should be the map view itself.

I found it hard to find the screen point from a feature. I had to do the following

```swift
let point = feature.geometry?.extent.center
let sourcePoint = mapView.location(toScreen: point!)
let sourceRect = CGRect(origin: sourcePoint, size: .zero)
```

to get the source rect of a feature. Is it the right way to go?

---

My intuition was to point the popover's arrow to the accessory button of the callout, since the button's tap action initiated the popover's presentation. It wasn't easy to find the frame of the accessory button, so I hacked to anchor the arrow to the callout's frame instead. See the top screenshot.

![image](https://user-images.githubusercontent.com/9660181/92024057-7ec11a00-ed12-11ea-8786-38d7d4164060.png)

Reply:
@yo1995 I'm not seeing the callout when the action sheet is displayed.

Reply:
I was using an iPad. The issue is the the action sheet covers the callout:

![image](https://user-images.githubusercontent.com/2257493/92027626-ab2b6500-ed17-11ea-9a1d-22823d1366a3.png)

Reply:
I see. It definitely looks better with arrow pointing at the feature.